src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+            [NotNullWhen(true)]  out ITypeSymbol? targetNodeType, out ImmutableArray<ITypeSymbol> potentialConversionTypes)
+        {
+            potentialConversionTypes = ImmutableArray<ITypeSymbol>.Empty;
+
+            var targetNodeInfo = semanticModel.GetTypeInfo(targetNode, cancellationToken);
+            targetNodeType = targetNodeInfo.Type;
+
+            if (targetNodeType == null)
+            {
+                return false;
+            }
+
+            // The error happens either on an assignement operation or on an invocation expression.
+            // If the error happens on assignment operation, "ConvertedType" is different from the current "Type"
+            var mutablePotentialConversionTypes = ArrayBuilder<ITypeSymbol>.GetInstance();
+            if (diagnosticId == "CS0266" && targetNodeInfo.ConvertedType != null && !targetNodeType.Equals(targetNodeInfo.ConvertedType))
+            {
+                mutablePotentialConversionTypes.Add(targetNodeInfo.ConvertedType);
+            }
+            else if (diagnosticId == "CS1503" && targetNode.GetAncestorsOrThis<ArgumentSyntax>().FirstOrDefault() is ArgumentSyntax targetArgument
+                && targetArgument.Parent is ArgumentListSyntax argumentList
+                && argumentList.Parent is SyntaxNode invocationNode) // invocation node could be Invocation Expression, Object Creation, Base Constructor...
+            {
+                // Implicit downcast appears on the argument of invocation node, get all candidate functions and extract potential conversion types 
+                var symbolInfo = semanticModel.GetSymbolInfo(invocationNode, cancellationToken);
+                var candidateSymbols = symbolInfo.CandidateSymbols;
+
+                foreach (var candidateSymbol in candidateSymbols)
+                {
+                    if (!(candidateSymbol is IMethodSymbol methodSymbol))
+                    {
+                        continue;
+                    }
+
+                    if (IsArgumentListAndParameterListPerfectMatch(semanticModel, argumentList.Arguments, methodSymbol.Parameters, targetArgument, cancellationToken, out var paramIndex))
+                    {
+                        var correspondingParameter = methodSymbol.Parameters[paramIndex];
+                        var argumentConversionType = correspondingParameter.Type;
+
+                        if (correspondingParameter.IsParams && correspondingParameter.Type is IArrayTypeSymbol arrayType && !(targetNodeType is IArrayTypeSymbol))
+                        {
+                            // target argument is matched to the parameter with keyword params
+                            argumentConversionType = arrayType.ElementType;
+                        }
+
+                        mutablePotentialConversionTypes.Add(argumentConversionType);
+                    }
+                }
+
+                // Sort the potential conversion types by inheritance distance
```

Comment:
this is an example of a comment that says 'what' you are doing (and just repeats the code below), but not 'why'.  I can tell you're sorting the potential conversion types by distance, because that's exactly what the code says.  What i can't tell is: what does this mean?  And why do we want to do it?

A good comment will explain that bit :)

Reply:
sure :)